src/librustdoc/passes/doc_test_lints.rs

```diff
@@ -89,10 +90,22 @@ pub fn look_for_tests<'tcx>(cx: &DocContext<'tcx>, dox: &str, item: &Item) {
 
     find_testable_code(&dox, &mut tests, ErrorCodes::No, false, None);
 
+    if cx.render_options.document_private
+        && dox.is_empty()
+        && !matches!(item.inner, clean::ImportItem(..))
+    {
+        let sp = span_of_attrs(&item.attrs).unwrap_or(item.source.span());
+        if !sp.is_dummy() {
+            cx.tcx.struct_span_lint_hir(rustc_lint::builtin::MISSING_DOCS, hir_id, sp, |lint| {
+                lint.build("missing documentation").emit()
+            });
+        }
+    }
```

Comment:
I'm not sure about this change.

- `missing_docs` is a rustc pass, not a `rustdoc` pass. I don't like that both rustc and rustdoc will run the same pass but _with different diagnostics_.
- clippy already lints for missing docs in private items (`missing_docs_in_private_items`). Do we really need to duplicate that here?

Reply:
It's an "outside" tool. Also, `--document-private-items` is a rustdoc option, meaning you can't check it in the rustc pass. This is more an extension than a pass in itself considering that it's on top.

Reply:
Right, but my concern is that this doesn't _belong_ in rustdoc. At least I'd want an FCP to be sure the rest of the team is agrees, if you feel strongly about it go ahead and start one.

Reply:
An FCP seems a bit big for this. Let's ping others, it might just be simpler.

What do you think about this @Manishearth, @ollie27 ?

Reply:
Yeah I do not like that the pass is duplicated with different diagnostics.

Clippy is an outside tool as much as rustdoc is :)

Reply:
But clippy isn't installed by default, I consider it "tier 2" in this regard. My point was mostly that it seems definitely like a good thing to have this check performed in rustdoc directly since it's a rustdoc option. However, I'm curious on how clippy does this check without having access to the rustdoc option.

Reply:
Yes, it is installed by default (and has been for a year now).

Clippy's check is independent of the rustdoc option, clippy's check is a lint you can enable and it will always check your private items. It's a separate lint.

To be clear, I don't think the clippy thing is anything more than a stop gap for this issue, but what I'm trying to say is: if we are duplicating the lint in rustdoc, it should behave _exactly_ the same way as the rustc one modulo private item behavior. This PR is inadequate: what we need to do is have a way for rustc's settings object to know about document-private-items, so when rustdoc internally constructs a compiler instance it can be set and the missing-docs lint can consume it.

Reply:
I don't like the idea of lints changing behavior based on `--document-private-items`. I think a separate lint to for private items would be the right solution here. Such a lint already exists in clippy so the only possible next step would be to move that lint to rustc.

Reply:
@Manishearth Wo, I didn't that about clippy, that's great! Also, I think you're misunderstanding something: it is not *in place*, it is an extension. The original lint check is still run (before this one in fact). So it's definitely not a duplication.

@ollie27: It doesn't change behavior, it's just that if you're documenting private items, it doesn't make sense to complain about the fact that they're private. And there is no way to know from rustc that `--document-private-items` has been passed to rustdoc (well, unless we add it to rustc directly obviously, but I'd rather notà.

Reply:
> The original lint check is still run (before this one in fact). So it's definitely not a duplication

Right, but the diagnostics are different. The code for this should be in one place.

> Such a lint already exists in clippy so the only possible next step would be to move that lint to rustc.

Totally happy to do this too. I'm not a huge fan of the lint changing behavior either.

Reply:
I have no issue putting this code in rustc directly, however it means that rustc will need to be aware of the `--document-private-items` rustdoc setting.

Reply:
No, i'm saying that we can _make_ rustc aware of that setting, by making it an internal flag that can be swapped by consumers of the rustc API (like rustdoc)

Reply:
I'm not sure we _should_ make rustc aware of it. I'd expect this to be a separate pass: `--document-private` should only affect whether the docs are _rendered_, not whether they're processed at all. `Strip` makes me uncomfortable for that reason, I'd prefer rustdoc linted consistently.